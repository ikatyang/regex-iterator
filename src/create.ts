/**
 * create a generator for string matching
 */
export function* create(regex: RegExp, str: string, max_result = Infinity) {
  const cloned_regex = new RegExp(regex.source, regex.flags);

  let counter = 0;
  let result = cloned_regex.exec(str);

  while (counter < max_result && result !== null) {
    counter++;

    yield result;

    if (!regex.global) {
      break;
    }

    result = cloned_regex.exec(str);
  }
}
