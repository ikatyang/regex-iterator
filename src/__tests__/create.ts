import {create} from '../create';

let regex: RegExp;
let global_regex: RegExp;
const str = '123 456 789';

beforeEach(() => {
  regex = /[0-9]+/;
  global_regex = /[0-9]+/g;
});

it('should clone the regex to exec', () => {
  const generator = create(global_regex, str);

  const pre_last_index = global_regex.lastIndex;
  generator.next();
  const post_last_index = global_regex.lastIndex;

  expect(pre_last_index).toBe(post_last_index);
});

it('should return correctly with global regex', () => {
  const generator = create(global_regex, str);
  expect([...generator]).toMatchSnapshot();
});

it('should return correctly without global regex', () => {
  const generator = create(regex, str);
  expect([...generator]).toMatchSnapshot();
});

it('should return correctly with max_result', () => {
  const max_result = 2;
  const generator = create(global_regex, str, max_result);
  expect([...generator].length).toBe(max_result);
});
