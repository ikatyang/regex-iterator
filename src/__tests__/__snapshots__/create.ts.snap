// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly with global regex 1`] = `
Array [
  Array [
    "123",
  ],
  Array [
    "456",
  ],
  Array [
    "789",
  ],
]
`;

exports[`should return correctly without global regex 1`] = `
Array [
  Array [
    "123",
  ],
]
`;
